evm/src/executor/backend/mod.rs

```diff
@@ -1665,22 +1680,34 @@ fn is_contract_in_state(journaled_state: &JournaledState, acc: Address) -> bool
 }
 
 /// Executes the given transaction and commits state changes to the database _and_ the journaled
-/// state
+/// state, with an optional inspector
 fn commit_transaction(
     tx: Transaction,
     mut env: Env,
     journaled_state: &mut JournaledState,
     fork: &mut Fork,
+    fork_id: &ForkId,
+    cheatcodes_inspector: Option<&mut Cheatcodes>,
 ) {
     configure_tx_env(&mut env, &tx);
-    let (_, state) = {
+
+    let (result, state) = {
         let mut evm = EVM::new();
         evm.env = env;
-        evm.database(&mut fork.db);
-        evm.transact()
+
+        let db = Backend::new_with_fork(fork_id, fork.clone(), journaled_state.clone());
+        evm.database(db);
+
+        if let Some(inspector) = cheatcodes_inspector {
+            evm.inspect(inspector)
+        } else {
+            evm.transact()
+        }
     };
 
-    apply_state_changeset(state, journaled_state, fork);
+    if let return_ok!() = result.exit_reason {
+        apply_state_changeset(state, journaled_state, fork);
```

Comment:
is this necessary? changeset should be reverted if transaction reverts, but should include balance changes for the sender I think